---
params:
  workingDir    : "in/"
  reportTitle   : "MultiGWAS report"
  nBest         : 6

output: 
  html_document: 
    theme: cerulean 
    toc: true
    number_sections: true
---

---
title  : `r params$reportTitle`

---

This report, created by the MultiGWAS tool, is a summary of the input parameters used to run the tool and its main outputs. The outputs include the best-ranked and significative SNPs found for each tool. For each one, there is a score table, Venn diagram, and an SNP profile.
 
```{r includeKnit, include=FALSE,echo=FALSE}
library(knitr)
library(kableExtra)
```

```{r  echo=FALSE, message=FALSE}
inputDir = paste0 (params$workingDir, "/report/")
nBest    = params$nBest
```

```{r setKnitOptions, include=FALSE}
knitr::opts_chunk$set(fig.width=12, fig.height=8, echo=FALSE, warning=FALSE, message=FALSE)
cachedata = FALSE
cachemodels = FALSE
```

# Input Parameters

MultiGWAS uses as input a configuration file where the user specifies the input genomic data (genotype and phenotype files) along with different values for parameters used by the MultiGWAS tool and by the other four GWAS
software (GWASpoly, SHEsis, PLINK, and TASSEL).

The following table shows the current input parameters specified by the user in the configuration file:

```{r printTableInputParameters }
paramsDF = read.table (file=paste0(inputDir,"out-multiGWAS-inputParameters.tbl"), header=T, sep="\t")
kable (paramsDF) %>%
 	kable_styling(bootstrap_options = c("striped", "hover", "condensed", font_size=5))
```

# Best-ranked SNPs
This section shows a table and a Venn diagram for the best ranked SNPs (For this report, N=`r nBest`). The configuration file allows defining the parameter N. 

## Table of best-ranked SNPs
The following table shows the associations between SNPs and the trait as transformed p-values to scores as -log 10 (p-value) sorted from the highest to the lowest scores based on each tool analysis and the model selected (Full or Naive). Each SNP has its associated information, such as the chromosome, position in the genome, and name. Also, based on the threshold calculated by a multiple testing correction (i.e., Bonferroni or FDR), the table marks the significance of each SNP as a TRUE or FALSE.


```{r printTableBest,  echo=FALSE, message=FALSE}
bestTable = read.table (file=paste0(inputDir, "/out-multiGWAS-scoresTable-best.scores"), header=T, sep="\t")
kable (bestTable) %>%
 	kable_styling(bootstrap_options = c("striped", "hover", "condensed", font_size=5))
```

## Venn diagram of best-ranked SNPs

The Venn diagram shows the best-ranked SNPs with high scores that are either unique or shared by the four software. Shared SNPs appear in the intersection areas of the diagram.

```{r vennDiagramBest, fig.align = "center", echo=FALSE, message=FALSE, fig.width=5, fig.height=5, fig.fullwidth=TRUE}
knitr::include_graphics (paste0(inputDir,"/out-multiGWAS-vennDiagram-best.png"))
```


# Significative SNPs
This section shows a table and a Venn diagram for the signifcative SNPs (score is above the significance threshold for each tool).  


## Table of significative SNPs

The following table shows the associations between SNPs and the trait as transformed p-values to scores as -log 10 (p-value) sorted from the highest to the lowest scores based on each tool analysis and the model selected (Full or Naive). Each SNP has its associated information, such as the chromosome, position in the genome, and name. 


```{r  echo=FALSE, message=FALSE}
significativesTable = read.table (file=paste0(inputDir, "/out-multiGWAS-scoresTable-significatives.scores"), header=T, sep="\t")
kable (significativesTable) %>%
 	kable_styling(bootstrap_options = c("striped", "hover", "condensed", font_size=5))
```


## Venn diagram of significative SNPs

The Venn diagram shows the significative SNPs that are either unique or shared by the four software. Shared SNPs appear in the intersection areas of the diagram.

```{r, fig.align = "center", echo=FALSE, message=FALSE, fig.width=5, fig.height=5, fig.fullwidth=TRUE}
knitr::include_graphics (paste0 (inputDir,"/out-multiGWAS-vennDiagram-significatives.png"))
```


# Manhattan and QQ plots
MultiGWAS uses classical Manhattan and Quantile–Quantile (QQ) plots to visually summarize GWAS results and identify both best-ranked and significative SNPs for each GWAS software. The Manhattan plot shows each SNPs distributed in the genetic location (x-axis) vs. the p-value (y-axis). The best-ranked SNPs are above the blue line and the significative SNPs above the blue line. In green are the SNPs that got high scores in more than one software. The QQ plot plots the observed (black dot lines) vs. expected -log 10 (p-value) (red line).

```{r  fig.widht=11, fig.height=30, fig.fullwidth=TRUE}
knitr::include_graphics (paste0 (inputDir, "/out-multiGWAS-manhattanQQ-plots.png"))
```

# Profiles for common significative SNPs 
For the common significative SNPs, MultiGWAS provides a figure to visualize each trait by individuals (rows) and genotypes (columns). At the left, the individuals are grouped in a dendrogram by their genotype. At the right, there is the name or ID of each individual. At the bottom, the genotypes are ordered from left to right, starting from the major to the minor allele (i.e., AAAA, AAAB, AABB, ABBB, BBBB). At the top, there is a description of the trait based on a histogram of frequency (top left) and by an assigned color for each numerical phenotype value using a grayscale (top right). Thus, each individual appears as a colored line by its phenotype value on its genotype column. For each column, there is a solid cyan line with the mean of each column and a broken cyan line that indicates how far the cell deviates from the mean.

Because each multiGWAS report shows one specific trait at a time, the histogram and color key will remain the same for all the best-ranked SNPs.

```{r, results='asis', echo=FALSE}
snpFileList =  list.files(inputDir, pattern=paste0("^(.*(","SNPProfile",").*(png)[^$]*)$"), full.names=T)
for (snpFile in snpFileList) {
	cat("![](", snpFile, ")")
	cat ("\n")
}

```

