<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"  
  "http://www.w3.org/TR/html4/loose.dtd">  
<html > 
<head><title>MultiGWAS: A software for GWAS analysis on
tetraploid organisms by integrating results of four GWAS
tools</title> 
<meta http-equiv="Content-Type" content="text/html; charset=windows-1252"> 
<meta name="generator" content="TeX4ht (http://www.tug.org/tex4ht/)"> 
<meta name="originator" content="TeX4ht (http://www.tug.org/tex4ht/)"> 
<!-- html,word,fn-in --> 
<meta name="src" content="multiGWAS-r4-doc.tex"> 
<link rel="stylesheet" type="text/css" href="multiGWAS-r4-doc.css"> 
</head><body 
>
                                             
                                                                                         

                                                                                         
<div class="maketitle">

<h2 class="titleHead">MultiGWAS: A software for GWAS analysis on tetraploid
organisms by integrating results of four GWAS tools</h2>
<div class="author" ></div><br />
<div class="date" ><span 
class="bchr8t-x-x-120">February 7, 2020</span></div>
</div>
<span 
class="bchb8t-x-x-80">Motivation: </span><span 
class="bchr8t-x-x-80">Although the GWAS software for polyploids organisms is scarce, other diploid software has also been use to analyze</span>
<span 
class="bchr8t-x-x-80">these organisms. This set of available GWAS software provides the possibility to replicate a study by using multiple tools, which</span>
<span 
class="bchr8t-x-x-80">helps to the researches to check for true associations when these ones are found by more than one tool. However, each tool</span>
<span 
class="bchr8t-x-x-80">has its own characteristics (interface, inputs, outputs, and arguments) which can may cost time and effort to successfully use</span>
<span 
class="bchr8t-x-x-80">them.</span><br 
class="newline" /><span 
class="bchb8t-x-x-80">Results: </span><span 
class="bchr8t-x-x-80">We have developed an R tool, MultiGWAS, to do GWAS analysis in polyploids organisms by executing and integrating the analysis results</span>
<span 
class="bchr8t-x-x-80">from four existing GWAS software: two for polyploids (GWASpoly and SHEsis) and the other two for diploids (Plink and Tassel). MultiGWAS</span>
<span 
class="bchr8t-x-x-80">allows users to run two kinds of GWAS: Full and Naive. The first considering the population structure and cryptic relatedness, and the second</span>
<span 
class="bchr8t-x-x-80">without these considerations. MultiGWAS receives a simple configuration file and the tool deals with all matters of the GWAS process to</span>
<span 
class="bchr8t-x-x-80">finally summarize and report results in a table and intuitive graphics formats to help researches to check potential true or false</span>
<span 
class="bchr8t-x-x-80">associations.</span><br 
class="newline" /><span 
class="bchb8t-x-x-80">Availability and implementation: </span><span 
class="bchr8t-x-x-80">Source code is freely available at https://github.com/agrosavia/multiGWAS along with examples,</span>
<span 
class="bchr8t-x-x-80">documentation and installation instructions.</span><br 
class="newline" /><span 
class="bchb8t-x-x-80">Contact: </span><span 
class="bchr8t-x-x-80">lgarreta@agrosavia.co</span>
<!--l. 71--><p class="indent" >  <hr class="figure"><div class="figure" 
>
                                             
                                                                                         

                                                                                         
<a 
 id="x1-21"></a>
                                             
                                                                                         

                                                                                         
<!--l. 73--><p class="noindent" > <img 
src="multi-gwas-flowchart-horizontal-stages.png" alt="PIC"  
>
<br /> <div class="caption" 
><span class="id">Figure 1: </span><span  
class="content"><span 
class="bchr8t-x-x-80">Flowchart of the central steps in the MultiGWAS tool.</span>
<span 
class="bchr8t-x-x-80">The  input  is  a  simple  configuration  file  including  among  other</span>
<span 
class="bchr8t-x-x-80">information  the  genotype/phenotype  files,  which  are  cleaned,</span>
<span 
class="bchr8t-x-x-80">filtered, and converted to the specific formats for each tool. Results</span>
<span 
class="bchr8t-x-x-80">for the four tools are processed and summarized in score tables,</span>
<span 
class="bchr8t-x-x-80">charts showing shared and individual markers, SNPs profiles and</span>
<span 
class="bchr8t-x-x-80">Manhattan/QQ plots. </span></span></div><!--tex4ht:label?: x1-21 -->
                                             
                                                                                         

                                                                                         
<!--l. 81--><p class="indent" >  </div><hr class="endfigure">
<!--l. 83--><p class="indent" >  <hr class="figure"><div class="figure" 
>
                                             
                                                                                         

                                                                                         
<a 
 id="x1-32"></a>
                                             
                                                                                         

                                                                                         
<!--l. 84--><p class="noindent" ><div class="minipage"><div class="minipage">Case 1:
</div>  <img 
src="out-Gwaspoly-Naive-Bonf-Sign.png" alt="PIC"  
>
</div>
<!--l. 90--><p class="noindent" ><div class="minipage"><div class="minipage">Case 2:
</div>  <img 
src="out-Gwaspoly-Structure-Bonf-Best5.png" alt="PIC"  
></div>
<br /> <div class="caption" 
><span class="id">Figure 2: </span><span  
class="content"><span 
class="bchr8t-x-x-80">Marker plots generated by the MultiGWAS tool for</span>
<span 
class="bchr8t-x-x-80">the SolCAP potato panel. Case 1: Naive GWAS analysis, the four</span>
<span 
class="bchr8t-x-x-80">tools found the same set of five markers (center area, red marker</span>
<span 
class="bchr8t-x-x-80">names). Case 2, Full GWAS analysis, the two diploid tools found</span>
<span 
class="bchr8t-x-x-80">one common marker (upper-central area, red marker name), but</span>
<span 
class="bchr8t-x-x-80">the other two polyploid tools found a different common marker</span>
<span 
class="bchr8t-x-x-80">(lower-central  area,  yellow  marker  name).  In  both  cases,  other</span>
<span 
class="bchr8t-x-x-80">marker were found either by two tools or by one tool. </span></span></div><!--tex4ht:label?: x1-32 -->
                                             
                                                                                         

                                                                                         
<!--l. 104--><p class="indent" >  </div><hr class="endfigure">
<a 
 id="x1-4r1"></a>
<h3 class="sectionHead"><span class="titlemark">1    </span> <a 
 id="x1-10001"></a>Introduction</h3>
<!--l. 109--><p class="noindent" >One way to validate GWAS results from a study is to
replicate the analysis using a different tool, but using
different GWAS software in plants to perform GWAS
analyzes to the same genomic data often produce
different results. Although the underlying model
implemented in this kind of software could be different,
each software has its own assumptions which could have
a strong influence on the results which produces
different associations and confuses the researchers from
the true associations. Furthermore, many important
crops are polyploids and new genomic data of polyploid
plants are available much more frequently, but most
of the GWAS software to analyze them have been
developed for diploid organisms. Currently, two software
tailored for polyploid organisms are the R package
GWASpoly <a 
href="#XRosyara2016">Rosyara et al.</a> [<a 
href="#XRosyara2016">2016</a>] and the SHEsis
tool <a 
href="#XShen2016">Shen et al.</a> [<a 
href="#XShen2016">2016</a>], and widely used diploid
GWAS software, as Plink <a 
href="#XPurcell2007">Purcell et al.</a> [<a 
href="#XPurcell2007">2007</a>] and
Tassel <a 
href="#XBradbury2007">Bradbury et al.</a> [<a 
href="#XBradbury2007">2007</a>], have been also used to
analyze polyploids organisms by «diploidizing» the
marker data <a 
href="#XLindqvist2014">Lindqvist-Kreuze et al.</a> [<a 
href="#XLindqvist2014">2014</a>], <a 
href="#XShulz2016">Schulz
et al.</a> [<a 
href="#XShulz2016">2016</a>].
<!--l. 126--><p class="indent" >  However, each tool has its own way of doing things:
different user interfaces (GUI or command line based),
different input formats (phenotype and genotype
formats), different models and algorithm assumptions,
and different outputs. Consequently, in light of all these
considerations, researchers must spend great effort when
they try to replicate GWAS results by using different
tools. With this in mind, we developed the MultiGWAS
tool that with a single input it performs GWAS analyses
by using four different GWAS software in such a way to
help researchers in the selection of true associations
across GWAS. The tool scans the different results files,
calculates new scores and thresholds from resulting
markers, select best ranked and significative makers, and
create summary tables and plots that show in an
intuitive manner the common and different markers
found by the four tools.
<a 
 id="x1-1001r3"></a>
<h3 class="sectionHead"><span class="titlemark">2    </span> <a 
 id="x1-20002"></a>Methods and Implementation</h3>
<!--l. 143--><p class="noindent" >MultiGWAS was implemented in R. Genomic data files
and parameters are specified in a configuration file (seesection <a 
href="#x1-30001">2.1<!--tex4ht:ref: subsec:Adjust --></a>) which is passed as unique argument to
 MultiGWAS through the command-line interface using a
 terminal in Linux or Mac systems.
 <!--l. 148--><p class="indent" >    A flowchart of the main central steps involved in the
 three phases of the MultiGWAS tool is outlined in figure
 <a 
href="#x1-21">1<!--tex4ht:ref: fig:Flowchar --></a>. The first phase preprocesses the inputs; the second
 phase performs GWAS on each tool; and the third phase
 postprocesses the outputs to summarize the results.
 These phases are detailed below.
 <a 
 id="x1-2001r1"></a>
 <h4 class="subsectionHead"><span class="titlemark">2.1    </span> <a 
 id="x1-30001"></a>Adjustment phase</h4>
 <!--l. 156--><p class="noindent" >MultiGWAS takes as input a configuration file where
 the user specify the genomic data for the GWAS
 analysis (genotype and phenotype files), along with the
 arguments used by the four tools to perform GWAS
 quality control (QC) on this data (see below). It starts
 with a cleaning and filtering steps performed on the
 genomic data. The first to get the common individuals
 to work with, and the second, to exclude from the
 analysis individuals and SNPs that are likely to be of
 poor-quality. Then, the filtered genotype and phenotype
 are transformed to the specific formats required for each
 tool.
 <!--l. 166--><p class="indent" >    The used format of the phenotype and genotype are
 the same as the the GWASpoly software <a 
href="#XRosyara2016">Rosyara
 et al.</a> [<a 
href="#XRosyara2016">2016</a>], except that the phenotype file must
 include only one trait without covariates. For the
 genotype file, the first three column correspond to the
 marker name, chromosome number, and position in the
 chromosome; while the next columns contain the marker
 data for each individual in the population codified in the
 ”ACGT” format (e.g. AATT, CCGG, AAAT, GGCG). For
 the phenotype file, the first column contains the
 individual name, and the second contains the trait
 value.
 <!--l. 176--><p class="indent" >    The current arguments included in the configuration
 file are: (1) The genotype/phenotype filenames, (2) The
 genome-wide significance threshold <span 
class="cmmi-10"><font face="Symbol">a</font> </span>(commonly
 0.01 or 0.05), (3) The method for multiple testing
 correction (”Bonferroni” or ”FDR”), (4) The type
 of GWAS (”NaiveGWAS” or ”FullGWAS”), (6) The
 maximum number of best ranked association to be
 reported, and (6) TRUE or FALSE whether to use quality
 control (QC) filters or not. The QC filters used by
 MultiGWAS are: (a) MAF (minor allele frequency), (b)
                                                                                         

                                                                                         
MIND (individual missing rate), (c) GENO (SNP missing
rate), and (d) HWE (Hardy-Weinberg threshold).
<a 
 id="x1-3001r5"></a>
<h4 class="subsectionHead"><span class="titlemark">2.2    </span> <a 
 id="x1-40002"></a>Multi analysis phase</h4>
<!--l. 190--><p class="noindent" >Two types of GWAS analyses can be conducted by
MultiGWAS: Full and Naive. The first with control for
population structure and individual relatedness; and the
second without any of these controls. Controlling
of population structure is based on the principal
components (PCs). Each software calculates the PCs and
uses the top ten as covariates in the association analysis,
while controlling of individual relatedness is based on
kinship matrices calculated by each software, in the case
of Tassel and GWASpoly, and calculated externally by the
king software <a 
href="#XManichaikul2010">Manichaikul et al.</a> [<a 
href="#XManichaikul2010">2010</a>], in the case of
Plink and SHEsis.
<!--l. 200--><p class="indent" >  The allowed format for the marker data is the ”ACGT”
configuration which is suitable for the polyploid tools
GWASpoly and SHEsis, but not for the diploid tools Plink
and Tassel. So, our MultiGWAS tool ”diploidizes” the
tetraploid marker data for these two tools by coding
each marker in two ways: all possible homozygous
genotypes (AAAA, CCCC, GGGG, TTTT), are coded with
two nucleotides (AA, CC, GG, and TT); and all possible
heterozygous genotypes (e.g. AAAT, ... ,CCCG) are coded
with the combination of their reference and alternate
alleles calculated from the tetraploid marker (e.g. AT, ...
,CG).
<!--l. 210--><p class="indent" >  When the filtered genotype and phenotype are ready
for the four tools, each tool is executed in parallel by
MultiGWAS using a parameterized script for each tool.
Each script is called using as arguments the filtered
genotype and phenotype along with the set of specified
values for the arguments described in section <a 
href="#x1-30001">2.1<!--tex4ht:ref: subsec:Adjust --></a>. If the
user has had experience with one of the four GWAS
tools, these scripts can be modified to improve the
analysis according to the specific tool syntax and
arguments.
<a 
 id="x1-4001r6"></a>
<h4 class="subsectionHead"><span class="titlemark">2.3    </span> <a 
 id="x1-50003"></a>Integration phase</h4>
<!--l. 224--><p class="noindent" >After the execution of the four tools by MultiGWAS,
their outputs are processed to create tables, marker
plots, and Manhattan/QQ plots that summarizes their
resulting associations. Scores and thresholds for
the markers are calculated taking into account: the    p-value of each marker, the user specified correction
  method and significance level <span 
class="cmmi-10"><font face="Symbol">a</font></span>. These values are
  calculated taking into account only the number of valid
  genotype calls (nonmissing genotype, phenotype, and
  covariates).
  <!--l. 233--><p class="indent" >    Two tables and marker plots are created: one for the
  significative associations which score is greater than the
  threshold, and other for the best <span 
class="bchri8t-">N </span>ranked associations,
  with <span 
class="bchri8t-">N </span>defined by the user in the configuration file. The
  Manhattan and QQ plots correspond to the associations
  found by the GWASpoly tool.
  <a 
 id="x1-5001r4"></a>
  <h3 class="sectionHead"><span class="titlemark">3    </span> <a 
 id="x1-60003"></a>Results and Discussion</h3>
  <!--l. 242--><p class="noindent" >Figure 2 presents two marker plots generated by
  MultiGWAS in the analysis of the genomic data for the
  Solanaceae Coordinated Agricultural Project (SolCAP)
  potato diversity panel, used to test the GWASpoly
  software <a 
href="#XRosyara2016">Rosyara et al.</a> [<a 
href="#XRosyara2016">2016</a>]. In the first case (upper
  plot on the figure <a 
href="#x1-32">2<!--tex4ht:ref: fig:Cases --></a>), it shows that the four tools found
  a large set of common markers as the GWAS type
  used was Naive, without any controls for population
  structure or individual relatedness. However, two
  markers: the c1_8019 and the c2_25471 are reported
  by Rosayra et al., the first as the most significative
  association, and second as the second best ranked
  association.
  <!--l. 253--><p class="indent" >    In the second case (lower plot on the figure <a 
href="#x1-32">2<!--tex4ht:ref: fig:Cases --></a>), it
  shows that the set of common markers is reduced, as
  the GWAS type used was Full GWAS with controls
  for population structure (10 PCs) and individual
  relatedness (kinship matrix). Now, the two polyploid
  tools GWASpoly and SHEsis found the common marker
  c1_8019, as in the Naive GWAS. But, the two diploid
  tools found a different common marker, the c1_16001.
  <a 
 id="x1-6001r8"></a>
  <h3 class="sectionHead"><span class="titlemark">4    </span> <a 
 id="x1-70004"></a>Conclusion</h3>
  <!--l. 263--><p class="noindent" >The MultiGWAS tool allows users to perform a GWAS
  analysis for tetraploid organism using four GWAS
  software at the same time: GWASpoly, SHEsis, Plink and
  Tassel. The tool deals with all the matters of the GWAS
  process in the four software, and only needs as input a
  simple configuration file with genomic filenames, values
  for different quality control filters, and the type of
  analysis to be performed. Analysis include both Full
  GWAS with control for population structure and
  individual relatedness, and Naive GWAS without any
                                                                                         

                                                                                         
control.
<!--l. 272--><p class="indent" >  The summary reports generated by the MultiGWAS
tool provide the user with tables and plots that show the
significative and best ranked markers found by each tool.
And, the marker plots is a powerful visualization that
describes graphically the markers found by each tool and
the common marker found by more than one tool, which
help users to decide in a more intuitively way the
possible true or false associations.
<a 
 id="Q1-1-10"></a>
<h3 class="likesectionHead"><a 
 id="x1-80004"></a>References</h3>
  <div class="thebibliography">
  <p class="bibitem" ><span class="biblabel">
<a 
 id="XBradbury2007"></a><span class="bibsp">   </span></span>P. J.  Bradbury,  Z. Zhang,  D. E.  Kroon,  T. M.
  Casstevens,    Y. Ramdoss,    and    E. S.    Buckler.
  TASSEL:   software   for   association   mapping   of
  complex traits in diverse samples.  <span 
class="bchri8t-">Bioinformatics</span>,
  23(19):2633—2635,   2007.       ISSN   1367-4803.
  doi:    10.1093/bioinformatics/btm308.          URL
  <a 
href="https://doi.org/10.1093/bioinformatics/btm308" class="url" ><span 
class="ectt-1000">https://doi.org/10.1093/bioinformatics/btm308</span></a>.
  </p>
  <p class="bibitem" ><span class="biblabel">
<a 
 id="XLindqvist2014"></a><span class="bibsp">   </span></span>H. Lindqvist-Kreuze,     M. Gastelo,     W. Perez,
  G. A.  Forbes,  D. De  Koeyer,  and  M. Bonierbale.
  Phenotypic stability and genome-wide association
  study of late blight resistance in potato genotypes
  adapted to the tropical highlands. <span 
class="bchri8t-">Phytopathology</span>,
  104(6):624—633,  2014.    ISSN  0031949X.    doi:
  10.1094/PHYTO-10-13-0270-R.
  </p>
  <p class="bibitem" ><span class="biblabel">
<a 
 id="XManichaikul2010"></a><span class="bibsp">   </span></span>A. Manichaikul,  J. C.  Mychaleckyj,  S. S.  Rich,
  K. Daly,   M. Sale,   and   W. M.   Chen.      Robust
  relationship inference in genome-wide association
  studies.  <span 
class="bchri8t-">Bioinformatics</span>, 26(22):2867—2873, 2010.
  ISSN  13674803.     doi:  10.1093/bioinformatics/
  btq559.
  </p>
  <p class="bibitem" ><span class="biblabel">
<a 
 id="XPurcell2007"></a><span class="bibsp">   </span></span>
  S. Purcell, B. Neale, K. Todd-Brown, L. Thomas,
  M. A. Ferreira, D. Bender, J. Maller, P. Sklar, P. I.
  De  Bakker,  M. J.  Daly,  and  P. C.  Sham.   PLINK:
  A   tool   set   for   whole-genome   association   and      population-based   linkage   analyses.      <span 
class="bchri8t-">American</span>
     <span 
class="bchri8t-">Journal of Human Genetics</span>, 81(3):559—575, 2007.
     ISSN 00029297. doi: 10.1086/519795.
     </p>
     <p class="bibitem" ><span class="biblabel">
<a 
 id="XRosyara2016"></a>  <span class="bibsp">   </span></span>U. R.  Rosyara,  W. S.  De  Jong,  D. S.  Douches,
     and J. B. Endelman.   Software for Genome-Wide
     Association   Studies   in   Autopolyploids   and   Its
     Application   to   Potato.        <span 
class="bchri8t-">The   Plant   Genome</span>,
     9(2):0,    2016.          ISSN    1940-3372.          doi:
     10.3835/plantgenome2015.08.0073.              URL
     <a 
href="https://dl.sciencesocieties.org/publications/tpg/abstracts/9/2/plantgenome2015.08.0073" class="url" ><span 
class="ectt-1000">https://dl.sciencesocieties.org/publications/tpg/abstracts/9/2/plantgenome2015.08.0073</span></a>.
     </p>
     <p class="bibitem" ><span class="biblabel">
<a 
 id="XShulz2016"></a>  <span class="bibsp">   </span></span>D. F. Schulz, R. T. Schott, R. E. Voorrips, M. J.
     Smulders,                     M. Linde,                     and
     T. Debener.     Genome-wide  association  analysis
     of   the   anthocyanin   and   carotenoid   contents
     of  rose  petals.      <span 
class="bchri8t-">Frontiers  in  Plant  Science</span>,  7
     (DECEMBER2016):1—15, 2016.   ISSN 1664462X.
     doi: 10.3389/fpls.2016.01798.
     </p>
     <p class="bibitem" ><span class="biblabel">
<a 
 id="XShen2016"></a>  <span class="bibsp">   </span></span>J. Shen,  Z. Li,  J. Chen,  Z. Song,  Z. Zhou,  and
     Y. Shi. SHEsisPlus, a toolset for genetic studies on
     polyploid species. <span 
class="bchri8t-">Scientific Reports</span>, 6:1—10, 2016.
     ISSN  20452322.   doi:  10.1038/srep24095.   URL
     <a 
href="http://dx.doi.org/10.1038/srep24095" class="url" ><span 
class="ectt-1000">http://dx.doi.org/10.1038/srep24095</span></a>.
  </p>
     </div>
      
</body></html> 

                                                                                         


