---
params:
  workingDir    : "in/"
  reportTitle   : "MultiGWAS report"

output: 
  html_document: 
    theme: cerulean 
    toc: true
    number_sections: true
---

---
title  : `r params$reportTitle`
---
This is the report created by the MultiGWAS tool that includes the input parameters used to run the tool along with its main outputs. The outputs include the best and significative SNPs found for each tool which are showed as Venn diagrams and score tables, and also include a particular visualization of the SNPs found by more than one tool that we have called as the "SNP profile" as it is a unique representation for a SNP. 

```{r includeKnit, include=FALSE,echo=FALSE}
library(knitr)
library(kableExtra)
```

```{r  echo=FALSE, message=FALSE}
inputDir     = paste0 (params$workingDir, "/report/")
```

```{r setKnitOptions, include=FALSE}
knitr::opts_chunk$set(fig.width=12, fig.height=8, echo=FALSE, warning=FALSE, message=FALSE)
cachedata = FALSE
cachemodels = FALSE
```

# Input Parameters
MultiGWAS uses as input a configuration file where the user specifies the input genomic data (genotype and phenotype files) along with different values for parameters used by the MultiGWAS tool and by the other four GWAS tools (GWASpoly, SHEsis, PLINK, and TASSEL). 

This section shows the current input parameters specified by the user in the configuration file:

```{r printTableInputParameters }
paramsDF = read.table (file=paste0(inputDir,"out-multiGWAS-inputParameters.tbl"), header=T, sep="\t")
kable (paramsDF) %>%
 	kable_styling(bootstrap_options = c("striped", "hover", "condensed", font_size=5))
```

\newpage
# Best Ranked SNPs
GWAS tools report the associations between SNPs and the traits as p-values. MultiGWAS transforms these p-values to scores as -log10(p-value), sorts them from high to low,  calculates a significance threshold according to the method used for multiple testing correction (Bonferroni or FDR), and marks each SNP as significative if their score is greater than the threshold.

Here, MultiGWAS report the best ranked SNPs for each tool using Venn diagrams and score tables. Venn diagrams show in an intuitive manner the best ranked SNPs shared by the four tools. Shared SNPs (high scores in more than one tool) appear in the intersection areas of the diagram, while individual SNPs (high scores only in a specific tool) appear in the ellipse of each tool. Score tables show more detailed information of each SNP as: the GWAS tool which scores the SNP, the GWAS model used by the tool (Full or Naive), the chromosome and its genetic location, the SNP name, the p-value and its score as -log10 (p-value), the significative threshold, and  a TRUE or FALSE flag wheter the SNP is significative or not. 


## Venn diagram of best ranked SNPs
```{r vennDiagramBest, fig.align = "center", echo=FALSE, message=FALSE, fig.width=5, fig.height=5, fig.fullwidth=TRUE}
knitr::include_graphics (paste0(inputDir,"/out-multiGWAS-vennDiagram-best.png"))
```

## Table of best ranked SNPs
```{r printTableBest,  echo=FALSE, message=FALSE}
bestTable = read.table (file=paste0(inputDir, "/out-multiGWAS-scoresTable-best.scores"), header=T, sep="\t")
kable (bestTable) %>%
 	kable_styling(bootstrap_options = c("striped", "hover", "condensed", font_size=5))
```

\newpage
# Significative SNPs
MultiGWAS shows the significative SNPs for each tool corresponding to the SNPs above the significance threshold (see above). As shown earlier, these SNPs are shown by two ways: a Venn diagram showing the shared SNPs, and a table showing the SNPs details.

## Venn diagram of signficative SNPs
```{r, fig.align = "center", echo=FALSE, message=FALSE, fig.width=5, fig.height=5, fig.fullwidth=TRUE}
knitr::include_graphics (paste0 (inputDir,"/out-multiGWAS-vennDiagram-significatives.png"))
```

## Table of significative SNPs
```{r  echo=FALSE, message=FALSE}
significativesTable = read.table (file=paste0(inputDir, "/out-multiGWAS-scoresTable-significatives.scores"), header=T, sep="\t")
kable (significativesTable) %>%
 	kable_styling(bootstrap_options = c("striped", "hover", "condensed", font_size=5))
```

\newpage
# Manhattan and QQ plots
MultiGWAS utilizes classical Manhattan and Quantileâ€“Quantile (QQ) plots to visually summarize GWAS results and identify both best ranked and significative SNPs. These plots are presented for each GWAS tool, the QQ plot plots the observed vs. expected -log10(p-value), and the Manhattan plot plots each of the SNPs as a dot with its genetic location on the x-axis and its p-value on the y-axis. 

Furthermore, MultiGWAS shows the following details in the Manhattan plots:

   - The best ranked SNPs, above the blue line.
   - The significative SNPs, above the red line.
   - The shared SNPs, marked in green (with high scores in more than one tool)

```{r  fig.widht=11, fig.height=30, fig.fullwidth=TRUE}
knitr::include_graphics (paste0 (inputDir, "/out-multiGWAS-manhattanQQ-plots.png"))
```

# Profiles for common significative SNPs 
For the common significative SNPs, MultiGWAS provides a figure to visualize each trait by individuals (rows) and genotypes (columns). At the left, the individuals are grouped in a dendrogram by their genotype. At the right, there is the name or ID of each individual. At the bottom, the genotypes are ordered from left to right, starting from the major to the minor allele (i.e., AAAA, AAAB, AABB, ABBB, BBBB). At the top, there is a description of the trait based on a histogram of frequency (top left) and by an assigned color for each numerical phenotype value using a grayscale (top right). Thus, each individual appears as a colored line by its phenotype value on its genotype column.

For each column, there is a solid cyan line with the mean of each column and a broken cyan line that indicates how far the cell deviates from the mean.

Because each multiGWAS report shows one specific trait at a time, the histogram and color key will remain the same for all the best-ranked SNPs.

```{r, results='asis', echo=FALSE}
snpFileList =  list.files(inputDir, pattern=paste0("^(.*(","SNPProfile",").*(png)[^$]*)$"), full.names=T)
for (snpFile in snpFileList) {
	cat("![](", snpFile, ")")
	cat ("\n")
}


```

